lib/spells/CDefaultSpellMechanics.cpp

```diff
@@ -445,118 +445,128 @@ void DefaultSpellMechanics::battleLogDefault(std::vector<MetaString> & logLines,
+	int duration = parameters.enchantPower;
+	//generate actual stack bonuses
 	{
-		SetStackEffect sse;
-		//get default spell duration (spell power with bonuses for heroes)
-		int duration = parameters.enchantPower;
-		//generate actual stack bonuses
-		{
-			int maxDuration = 0;
-			std::vector<Bonus> tmp;
-			owner->getEffects(tmp, parameters.effectLevel);
-			for(Bonus& b : tmp)
-			{
-				//use configured duration if present
-				if(b.turnsRemain == 0)
-					b.turnsRemain = duration;
-				vstd::amax(maxDuration, b.turnsRemain);
-				sse.effect.push_back(b);
-			}
-			//if all spell effects have special duration, use it
-			duration = maxDuration;
-		}
-		//fix to original config: shield should display damage reduction
-		if(owner->id == SpellID::SHIELD || owner->id == SpellID::AIR_SHIELD)
-		{
-			sse.effect.back().val = (100 - sse.effect.back().val);
-		}
-		//we need to know who cast Bind
-		if(owner->id == SpellID::BIND && parameters.casterStack)
-		{
-			sse.effect.back().additionalInfo =  parameters.casterStack->ID;
-		}
-		std::shared_ptr<Bonus> bonus = nullptr;
-		if(parameters.casterHero)
-			bonus = parameters.casterHero->getBonusLocalFirst(Selector::typeSubtype(Bonus::SPECIAL_PECULIAR_ENCHANT, owner->id));
-		//TODO does hero specialty should affects his stack casting spells?
+		si32 maxDuration = 0;
+
+		owner->getEffects(sse.effect, parameters.effectLevel, false, duration, &maxDuration);
+		owner->getEffects(sse.cumulativeEffects, parameters.effectLevel, true, duration, &maxDuration);
+
+		//if all spell effects have special duration, use it later for special bonuses
+		duration = maxDuration;
+	}
+	//fix to original config: shield should display damage reduction
+	if(owner->id == SpellID::SHIELD || owner->id == SpellID::AIR_SHIELD)
+	{
+		sse.effect.at(sse.effect.size()-1).val = (100 - sse.effect.back().val);
+	}
+	//we need to know who cast Bind
+	else if(owner->id == SpellID::BIND && parameters.casterStack)
+	{
+		sse.effect.at(sse.effect.size()-1).additionalInfo = parameters.casterStack->ID;
+	}
+	std::shared_ptr<Bonus> bonus = nullptr;
+	if(parameters.casterHero)
+		bonus = parameters.casterHero->getBonusLocalFirst(Selector::typeSubtype(Bonus::SPECIAL_PECULIAR_ENCHANT, owner->id));
+	//TODO does hero specialty should affects his stack casting spells?
 
+	for(const CStack * affected : ctx.attackedCres)
+	{
 		si32 power = 0;
-		for(const CStack * affected : ctx.attackedCres)
-		{
-			sse.stacks.push_back(affected->ID);
+		sse.stacks.push_back(affected->ID);
 
-			//Apply hero specials - peculiar enchants
-			const ui8 tier = std::max((ui8)1, affected->getCreature()->level); //don't divide by 0 for certain creatures (commanders, war machines)
-			if(bonus)
+		//Apply hero specials - peculiar enchants
+		const ui8 tier = std::max((ui8)1, affected->getCreature()->level); //don't divide by 0 for certain creatures (commanders, war machines)
+		if(bonus)
+		{
+			switch(bonus->additionalInfo)
```

Comment:
This switch also had formatting unlike all other in spells code.
Might be it's time to fix it if you change it anyway.